internal/experimental/pusher/pluginpusher.go

```diff
@@ -0,0 +1,102 @@
+/*
+Copyright The Helm Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package pusher
+
+import (
+	"bytes"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+
+	"github.com/pkg/errors"
+
+	"helm.sh/helm/v3/pkg/cli"
+	"helm.sh/helm/v3/pkg/plugin"
+)
+
+// collectPlugins scans for pusher plugins.
+// This will load plugins according to the cli.
+func collectPlugins(settings *cli.EnvSettings) (Providers, error) {
+	plugins, err := plugin.FindPlugins(settings.PluginsDirectory)
+	if err != nil {
+		return nil, err
+	}
+	var result Providers
+	for _, plugin := range plugins {
+		for _, uploader := range plugin.Metadata.Uploaders {
+			result = append(result, Provider{
+				Schemes: uploader.Protocols,
+				New: NewPluginPusher(
+					uploader.Command,
+					settings,
+					plugin.Metadata.Name,
+					plugin.Dir,
+				),
+			})
+		}
+	}
+	return result, nil
+}
+
+// pluginPusher is a generic type to invoke custom uploaders,
+// implemented in plugins.
+type pluginPusher struct {
+	command  string
+	settings *cli.EnvSettings
+	name     string
+	base     string
+	opts     options
+}
+
+// Push runs uploader plugin command
+func (p *pluginPusher) Push(chartRef, href string, options ...Option) error {
+	for _, opt := range options {
+		opt(&p.opts)
+	}
+	commands := strings.Split(p.command, " ")
+	argv := append(commands[1:], p.opts.certFile, p.opts.keyFile, p.opts.caFile, href, chartRef)
```

Comment:
You will notice this is almost identical to pluginGetter.Get (https://github.com/helm/helm/blob/main/pkg/getter/plugingetter.go#L66)

just adding the chart ref as an extra argument (so that uploaders can locate the chart which needs to be uploaded)